optimum/onnxruntime/modeling_decoder.py

```diff
@@ -630,14 +634,65 @@ def to(self, device: Union[torch.device, str, int]):
 
 
 @add_end_docstrings(ONNX_MODEL_END_DOCSTRING)
-class ORTModelForCausalLM(ORTModelDecoder, GenerationMixin):
+class ORTModelForCausalLM(ORTModel, GenerationMixin):
     """
     ONNX model with a causal language modeling head for ONNX Runtime inference. This class officially supports bloom, codegen, gpt2, gpt_bigcode, gpt_neo, gpt_neox, gptj, llama.
     """
 
     auto_model_class = AutoModelForCausalLM
     main_input_name = "input_ids"
 
+    def __init__(
+        self,
+        model: onnxruntime.InferenceSession,
+        config: "PretrainedConfig",
+        use_io_binding: Optional[bool] = None,
+        model_save_dir: Optional[Union[str, Path, TemporaryDirectory]] = None,
+        preprocessors: Optional[List] = None,
+        generation_config: Optional[GenerationConfig] = None,
+        use_cache: Optional[bool] = None,
+        **kwargs,
+    ):
+        if use_io_binding is None:
+            if model.get_providers()[0] in ["CPUExecutionProvider", "CUDAExecutionProvider"]:
+                use_io_binding = True
+            else:
+                use_io_binding = False
+
+        super().__init__(model, config, use_io_binding, model_save_dir, preprocessors, **kwargs)
+
+        self.num_pkv = 2
+        self.normalized_config = NormalizedConfigManager.get_normalized_config_class(config.model_type)(config)
+        self.key_value_input_names = [key for key in self.inputs_names if (".key" in key) or (".value" in key)]
+        self.key_value_output_names = [key for key in self.output_names if (".key" in key) or (".value" in key)]
+        self.use_cache = len(self.key_value_input_names) > 0
+
+        if generation_config is None:
+            generation_config = GenerationConfig.from_model_config(config)
+        self.generation_config = generation_config
+        self.onnx_paths = [self.model_path]
+        self.use_merged = "use_cache_branch" in self.inputs_names
+
+        self.use_fp16 = False
+        for inp in model.get_inputs():
+            if inp.name == "past_key_values" and inp.type == "tensor(float16)":
+                self.use_fp16 = True
+                break
+
+        if use_cache ^ self.use_cache:
```

Comment:
I personally find this clearer but both are correct.
```suggestion
        if use_cache != self.use_cache:
```